### ISSUE TITLE
Proposal: Auto-scaling capabilities based on container's metrics

### ISSUE THREAD

**Issue by cusspvz** [rancher/rancher#3315]
My proposal is to add an auto-scaling feature on cattle based on container's metrics.

Such could be configurable by:
- Network in
- Network out
- Network in/out
- CPU in use
- Memory in use

I know that such a thing could be a little bit heavy, so I also have a question: Do you have always access to containers metrics from cattle or just when it is requested by the UI?

**Comment by cusspvz** [comment]
I have a second in mind I would love to see, concerning hosts auto-scaling from within **Rancher**. In case you guys think it is a cool idea. :)

**Comment by RVN-BR** [comment]
It would be cool to have, nonetheless, I'd argue that perhaps the best would be a way to have service-specific metrics?  You could also run into issues if you dont limit your scaling with this sort of "auto scaling metric"... (if you have some code generating a memory leak for example, you can get to a condition where you overflow your hosts...)

I'd like to see something schedule based (in many workloads, schedule is the primary scaling factor. i.e. we know what typical traffic looks like, so we know to scale up at such a time - ex: 1h before traffic starts to go up, then we know to be at X hosts/containers at such a time, and we know when traffic drops out, etc)...  When CPU or Network is starting to become a limiting factor, it is usually very near the end (at least in our web app experiences)... But yes, as a secondary metric CPU or Memory are indeed useful...  As are metrics generated by the service such as active connections, etc...

**Comment by flypenguin** [comment]
so this is accepted and I can just wait for it? :grinning: Couse this would remove a big item for me.

**Comment by taketnoi** [comment]
Can't wait to see this feature in Rancher. ++1

**Comment by AKoetsier** [comment]
Autoscaling services would be really nice to have. K8s also has this and might serve as a source of inspiration/ideas on how it could be used within rancher? http://kubernetes.io/docs/user-guide/horizontal-pod-autoscaling/

**Comment by bonifaido** [comment]
I think what is possible here, and what already is done with AWS and GCE is to use the [kubernetes-contrib/cluster-autoscaler](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler) and create a generic docker-machine based cloudprovider there, which would allocate nodes based on a predefined configuration in Rancher.

**Comment by AKoetsier** [comment]
No what is described there is increasing the replication factor of the pods based on their CPU utilization. What you describe is scaling the instances on which the pods can be placed. So there is actually to types of scaling involved here, one is the scale setting of a service in rancher the other would be the hosts actually running the containers. Autoscaling the hosts can be done quit easily for instance with AWS by just using Cloudwatch, ASG triggers and some userdata. However scaling the services itself is not yet possible and that is what my link is referring to in k8s.

**Comment by lilnate22** [comment]
+1 for this. ideally the user should be able to launch couple of nodes based on some metrics...we can manually do this now, why not have it do it automatically

**Comment by ysle** [comment]
+1

**Comment by cusspvz** [comment]
> Autoscaling the hosts can be done quite easily for instance with AWS by just using Cloudwatch, ASG triggers and some userdata.

@AKoetsier yeah, most of providers give that power, but on some cases it may be not what you pretend and you just have to place ugly hacks on top of it, like launching the swarm agent. If there was a layer that were abstracted from the IaaS (which is, from my perspective, Rancher's milestone) and had rules by services it would be great.

**Comment by dsaydon90** [comment]
+1

**Comment by deniseschannon** [comment]
@cjellick can you comment on how the first round of autoscaling services will work in Rancher?

**Comment by johnculviner** [comment]
Heads up to the Rancher team FWIW:

I personally use Rancher quite a bit and give a Docker + Rancher talk in the Minneapolis. This is one of the main points of contention people have about using Rancher either during the talk or coming up afterwards. For my uses it doesn't matter much but it's definitely something people are thinking about. Only Con I could even come up with! :

![image](https://cloud.githubusercontent.com/assets/1644353/22389345/ab97f5a4-e4aa-11e6-8984-ca42f15688e5.png)

**Comment by abagshaw** [comment]
Any progress on this?

**Comment by cusspvz** [comment]
I'm also gonna give a talk next Wednesday, at [Microsoft Open-Source Camp](http://msoscamp.io/) and like @johnculviner, this is the only *cons* I will be showing up. :p

**Comment by ediri** [comment]
+1 

Same here could be nice to have an autoscaling feature in the cattle environment. 

Is this on some kind of roadmap?

**Comment by pmerson** [comment]
+1 
We're choosing between cattle and k8s in my organization. (We bought Rancher recently.) 
Autoscaling is one rather important criterion we're looking at. 
Is it planned in a near-term release?

**Comment by jonathanbaraldi** [comment]
I've done this. 

Just like that, just like Amazon Autoscaling.

**Comment by anjia0532** [comment]
@jonathanbaraldi  how to do ?

**Comment by jonathanbaraldi** [comment]
@anjia0532 
1) You need to create a Webhook inside Rancher. This WebHook is going to scale the service that you define. Select your service and define how much to scale up or down.
2) You need to use Prometheus and Grafana for monitoring cluster metrics.
3) You need to create a custom monitoring that keep looking to a cpu metrics or any other, filtering by the IMAGE of the container, NOT the name of the container.
4) You need to create an alert for that monitoring, like....if more than 50% cpu trigger the alert.
5) This alert is going to be a WebHook type, and is going to request that WebHook that you created in Rancher in the first step.
6) Done. With this, Grafana and prometheus are trigger Rancher Webhook to scale up or down, acoording to what you defined.

Is working very well, but I preferrer the K8S way. 
Better and dont need others softwares to work together.

**Comment by anjia0532** [comment]
@jonathanbaraldi Genius! thank you very much. i'll do it. Maybe I can use elasticsearch(like Prometheus)+elastalert(create custom monitoring rule and send alert)+webhook instead Prometheus+Grafana.

**Comment by gregkeys** [comment]
@jonathanbaraldi while using Prometheus is a great solution it doesn't scale well, we need a solution that we can bake into a rancher-compose and thus into a catalog, with Prometheus we have to add 30 webhooks for every application we bring online which doesn't seem maintainable at scale

**Comment by jonathanbaraldi** [comment]
Yes. This is a workaround.
Because of this I preferr to use Kubernetes insted of Cattle.
Kubernetes have native autoscaling capabilities that works with cpu, memory
and network.

You just simples add 6 or 7 lines in your template, and your application is
ready do scale.

I dont understand why docker-compose still dont have this. Not even Docker
Swarm or Cattle have this yet, its a shame.

I past the same situation in our company and with our customers.


Em 16 de ago de 2017 00:26, "Greg Keys" <notifications@github.com> escreveu:

@jonathanbaraldi <https://github.com/jonathanbaraldi> while using
Prometheus is a great solution it doesn't scale well, we need a solution
that we can bake into a rancher-compose and thus into a catalog, with
Prometheus we have to add 30 webhooks for every application we bring online
which doesn't seem maintainable at scale

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/rancher/rancher/issues/3315#issuecomment-322657128>,
or mute
the thread

**Comment by ranjithkumar-s** [comment]
@jonathanbaraldi How to dynamically add alerts rules in rules.conf and prometheus yml file via API or something in Promtheus. I have added prometheus stack for monitoring Rancher server. I have tried configuring alerts in prometheus stack by manually. It worked. Now i want to add the rules and configure webhooks in alertmanager config yml file dyanmically. Is there anyway I can do that?? Do they have API for this scenario?
